### PR TITLE
Issue/423

### PR DESCRIPTION
**Pull request by minborg**

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/hash/impl/util/CanonicalRandomAccessFiles.java [hunk 1]
```diff
@@ -50,7 +50,8 @@ private CanonicalRandomAccessFiles() {
     }
 
     public static RandomAccessFile acquire(@NotNull final File file) throws FileNotFoundException {
-        return acquire0(file, (ref) -> {}).raf;
+        return acquire0(file, (ref) -> {
```
Comment (minborg) [inline src/main/java/net/openhft/chronicle/hash/impl/util/CanonicalRandomAccessFiles.java]: Reformatted by IntelliJ. Should we revert?

Comment (alamar) [reply src/main/java/net/openhft/chronicle/hash/impl/util/CanonicalRandomAccessFiles.java]: Maybe we need some reusable empty consumer?

File: src/test/java/net/openhft/chronicle/map/issue/Issue423Test.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class Issue423Test {
```
Comment (alamar) [inline src/test/java/net/openhft/chronicle/map/issue/Issue423Test.java]: I suggest having a descriptive name. Issue is visible in commit message anyway

**Review by RobAustin — APPROVED** [review]

**Review by alamar — CHANGES_REQUESTED** [review]
Does this issue only trigger on Windows? I can't see the reproducer to fail. Otherwise, it makes sense.

File: src/main/java/net/openhft/chronicle/hash/impl/util/CanonicalRandomAccessFiles.java [hunk 3]
```diff
@@ -77,7 +78,8 @@ private static RafReference acquire0(@NotNull final File file, Consumer<RafRefer
     }
 
     public static void release(@NotNull final File file) {
-        release0(file, (ref) -> {});
+        release0(file, (ref) -> {
+        });
```
Comment (alamar) [inline src/main/java/net/openhft/chronicle/hash/impl/util/CanonicalRandomAccessFiles.java]: If you must keep it reformatted you could add `// No-op.` comment

**Review by alamar — CHANGES_REQUESTED** [review]

**Review by alamar — APPROVED** [review]

File: .gitignore [hunk 4]
```diff
@@ -55,3 +55,5 @@ local.properties
 *.cq4
 
 file.lock
+
+issue423
```
Comment (JerryShea) [inline .gitignore:59]: why?

File: src/test/java/net/openhft/chronicle/map/issue/Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+             FileChannel fileChannel = raf.getChannel();) {
+
+            try (FileLock lock = fileChannel.tryLock()) {
+                // Make sure we can lock (hence the file was not previously locked)
+                assertNotNull(lock);
+            }
+        }
+
+        // Make sure the file can be deleted despite an Exception was thrown by the builder
+        assertTrue(file.delete());
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/map/issue/Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest.java:43]: this should be in a finally?

File: src/test/java/net/openhft/chronicle/map/issue/Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import java.nio.channels.FileLock;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest {
+
+    @Test
+    public void reproduce() throws IOException {
+        final File file = new File("issue423");
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/map/issue/Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest.java:21]: maybe put in OS.target() so it does not clutter the directory

Comment (minborg) [reply src/test/java/net/openhft/chronicle/map/issue/Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest.java:21]: Good suggestion

File: src/test/java/net/openhft/chronicle/map/issue/Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest.java (hunk 5 above)
Comment (minborg) [reply src/test/java/net/openhft/chronicle/map/issue/Issue423LockFileRemainsLockedOnWindowsIfBuilderThrowsTest.java:43]: That would have been better, yes.

**Pull request merged**